src/passes/OptimizeAddedConstants.cpp

```diff
@@ -32,13 +32,35 @@
 
 #include <ir/local-graph.h>
 #include <ir/local-utils.h>
-#include <ir/parents.h>
 #include <pass.h>
 #include <wasm-builder.h>
 #include <wasm.h>
 
 namespace wasm {
 
+namespace {
+
+// Similar to Parents from parents.h, but we only care about gets, so it is much
+// more efficient to just collect their parents.
+struct GetParents {
+  GetParents(Expression* expr) { inner.walk(expr); }
+
+  Expression* getParent(LocalGet* curr) const {
+    auto iter = inner.parentMap.find(curr);
+    assert(iter != inner.parentMap.end());
+    return iter->second;
+  }
+
+private:
+  struct Inner : public ExpressionStackWalker<Inner> {
+    void visitLocalGet(LocalGet* curr) { parentMap[curr] = getParent(); }
+
+    std::map<Expression*, Expression*> parentMap;
```

Comment:
Looks like this can be `std::unordered_map`. I'd be curious if that makes a measurable difference.

Reply:
Good idea, it does help by 1%. Added.

Reply:
I'll separately see if the other class is likewise sped up with such a map.